### ISSUE TITLE
Moving to yarn 2

### ISSUE THREAD

**Issue by pixelastic** [pixelastic/aberlaas#90]
Yarn 2 has an impressive list of new features and bug fixing; many of them where patched internally by aberlaas

I should make yarn 2 the default of aberlaas and fix the related configuration

For example, the fact that ESLint/Jest plugins are loaded from the cwd and not the calling file, can be fixed like this: https://github.com/gatsbyjs/gatsby/pull/20629

**Comment by pixelastic** [comment]
I spent a fair amount of time today to try to move `firost` to Yarn 2, thinking it will help in some workspace issues I had. After being dragged pretty deep into the rabbit hole of issues, I think it's too early to switch.

Yarn 2 (codenamed `berry`) seems much faster than yarn. It seems to be using a local cache of the dependencies so installs seem to require less HTTP calls. I'm not 100% sure of that, but in my tests, installs seemed even faster than before.

Log output is clearer. Lines are colored so it's easy to spot module names and versions. Steps are groupes together and each line has a unique error code attached so we can refer to it in the documentation. From that POV, DX is much better

Dependencies are no longer saved in `node_modules`, they are in `yarn/cache`, but they are not exactly the same thing either. We shouldn't make any assumption on where a specific file is located (even for `.bin/`). Everything should pass through `yarn bin` and `yarn run`.

The above point is very interesting because it means that we should no longer have to guess where a file is located. The previous logic was non-deterministic, and based on if we were in a simple project, a mono-repo, using `yarn link` or global installs it was very hard to know where everything could be. We should not need to know where a binary is located, we should simply call it with `yarn run` and it will be found.

Dependencies are now strictly scoped. In a case of a mono-repo, your sub-packages cannot require modules that are dependencies of their siblings. In the previous implementation they could, because all dependencies where hoisted to the top level. This is no longer possible, and is a good thing overall.

In the same vein, you binaries of your sub-dependencies are no longer hoisted either. Only binaries of the direct dependencies are. This is where things get trickier. `aberlaas` makes calls to several binaries `eslint`, `stylelint`, etc. Those binaries can now only be called by their direct parent (ie, you can run `yarn run eslint` while in `aberlaas`, but this won't work when in a project that depends on `aberlaas`).

The current workaround of having to call `yarn run eslint` instead of the binary directly seemed to fail in my tests. Running `yarn run eslint` from inside aberlaas and running `yarn run aberlaas lint` from the root resulted in a process that never ended and took 100% of my CPU. I'm not sure where it's coming from.

Because we can no longer guess the exact path of the binaries, my custom vim helpers to automatically lint files on save where also confused and didn't lint, which decreased my own DX as well.

Not being able to call binaries of sub-dependencies also meant that it was not possible to directly call `eslint` from a project using `aberlaas` as it wasn't exposed. A workaround would have been to explicitly expose a binary named `eslint` from the `bin` field of the `package.json` of `aberlaas`. This would have been in theory cleaner as we know for sure which binaries are exported, but I couldn't find a way to call the underlying `eslint` from there. `yarn bin eslint` from there pointed to itself, completely shadowing the `eslint` inside the module (which was the point, but I couldn't find a way to bypass it).

**TL;DR;**

In a nutshell, `berry` removes `node_modules` and only exposes direct binary dependencies. This breaks any assumptions we had on calling binaries ourselves from `aberlaas` and broke many inclusion paths (plugins, local vim helpers, etc).

I think I should wait 6+ months before giving this another try.

**Solutions**

One solution would be to stop calling binaries from inside aberlaas and always use the node APIs of the tools. I'm not sure if this is possible as many of those tools have been thought as CLI-first and might not have an official node API surface to use.

**Comment by pixelastic** [comment]
Closing in favor of a clearer issue to explain the pain points and find a path to migration.
